Default database empty configuration is not working

- Create a new application with JPA support: `mn create-app foo --features=hibernate-jpa`
- According to the documentation (https://docs.micronaut.io/latest/guide/index.html#_configuring_jdbc_connection_pools) the following configuration:
```
datasources.default: {}
```

Is the same as:
```
datasources:
    default:
        url: jdbc:h2:mem:default;DB_CLOSE_DELAY=-1;DB_CLOSE_ON_EXIT=FALSE
        username: sa
        password: ""
        driverClassName: org.h2.Driver
```

Using the former configuration, the application fails when injecting `DataSource` with the error:
```
09:22:38.424 [main] INFO  io.micronaut.runtime.Micronaut - Startup completed in 1590ms. Server Running: http://localhost:8080
09:22:40.240 [nioEventLoopGroup-1-2] ERROR i.m.h.s.netty.RoutingInBoundHandler - Unexpected error occurred: Failed to inject value for parameter [dataSource] of class: foo.FooController

Message: No bean of type [javax.sql.DataSource] exists. Ensure the class is declared a bean and if you are using Java or Kotlin make sure you have enabled annotation processing.
Path Taken: new FooController([DataSource dataSource])
io.micronaut.context.exceptions.DependencyInjectionException: Failed to inject value for parameter [dataSource] of class: foo.FooController

Message: No bean of type [javax.sql.DataSource] exists. Ensure the class is declared a bean and if you are using Java or Kotlin make sure you have enabled annotation processing.
Path Taken: new FooController([DataSource dataSource])
	at io.micronaut.context.AbstractBeanDefinition.getBeanForConstructorArgument(AbstractBeanDefinition.java:981)
	at foo.$FooControllerDefinition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1303)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:1861)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:1597)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:1577)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:509)
	at io.micronaut.context.DefaultBeanContext$BeanExecutionHandle.getTarget(DefaultBeanContext.java:2333)
	at io.micronaut.context.DefaultBeanContext$BeanExecutionHandle.invoke(DefaultBeanContext.java:2354)
	at io.micronaut.web.router.AbstractRouteMatch.execute(AbstractRouteMatch.java:237)
	at io.micronaut.web.router.RouteMatch.execute(RouteMatch.java:123)
	at io.micronaut.http.server.netty.RoutingInBoundHandler.lambda$buildResultEmitter$14(RoutingInBoundHandler.java:1243)
	at io.reactivex.internal.operators.flowable.FlowableCreate.subscribeActual(FlowableCreate.java:71)
	at io.reactivex.Flowable.subscribe(Flowable.java:14479)
	at io.reactivex.internal.operators.flowable.FlowableMap.subscribeActual(FlowableMap.java:37)
	at io.reactivex.Flowable.subscribe(Flowable.java:14479)
	at io.reactivex.internal.operators.flowable.FlowableSwitchIfEmpty.subscribeActual(FlowableSwitchIfEmpty.java:32)
	at io.reactivex.Flowable.subscribe(Flowable.java:14479)
	at io.reactivex.Flowable.subscribe(Flowable.java:14429)
	at io.micronaut.http.context.ServerRequestTracingPublisher.lambda$subscribe$0(ServerRequestTracingPublisher.java:53)
	at io.micronaut.http.context.ServerRequestContext.with(ServerRequestContext.java:53)
	at io.micronaut.http.context.ServerRequestTracingPublisher.subscribe(ServerRequestTracingPublisher.java:53)
	at io.reactivex.internal.operators.flowable.FlowableFromPublisher.subscribeActual(FlowableFromPublisher.java:29)
	at io.reactivex.Flowable.subscribe(Flowable.java:14479)
	at io.reactivex.Flowable.subscribe(Flowable.java:14426)
	at io.reactivex.internal.operators.flowable.FlowableSubscribeOn$SubscribeOnSubscriber.run(FlowableSubscribeOn.java:82)
	at io.reactivex.internal.schedulers.ExecutorScheduler$ExecutorWorker$BooleanRunnable.run(ExecutorScheduler.java:260)
	at io.reactivex.internal.schedulers.ExecutorScheduler$ExecutorWorker.run(ExecutorScheduler.java:225)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: io.micronaut.context.exceptions.NoSuchBeanException: No bean of type [javax.sql.DataSource] exists. Ensure the class is declared a bean and if you are using Java or Kotlin make sure you have enabled annotation processing.
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:1583)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:886)
	at io.micronaut.context.AbstractBeanDefinition.getBeanForConstructorArgument(AbstractBeanDefinition.java:973)
	... 30 common frames omitted
```

This is related to #755

--- a/inject/src/main/java/io/micronaut/context/env/AbstractPropertySourceLoader.java
+++ b/inject/src/main/java/io/micronaut/context/env/AbstractPropertySourceLoader.java
@@ -139,7 +139,7 @@ public abstract class AbstractPropertySourceLoader implements PropertySourceLoad
             Map.Entry entry = (Map.Entry) o;
             String key = entry.getKey().toString();
             Object value = entry.getValue();
-            if (value instanceof Map) {
+            if (value instanceof Map && !((Map) value).isEmpty()) {
                 processMap(finalMap, (Map) value, prefix + key + '.');
             } else {
                 finalMap.put(prefix + key, value);

--- a/inject/src/test/groovy/io/micronaut/context/env/yaml/YamlPropertySourceLoaderSpec.groovy
+++ b/inject/src/test/groovy/io/micronaut/context/env/yaml/YamlPropertySourceLoaderSpec.groovy
@@ -80,4 +80,46 @@ dataSource:
         env.get("data-source.password", String).get() == 'test'
         env.get("data-source.jmx-export", boolean).get() == true
     }
+
+    void "test datasources.default: {}"() {
+        def serviceDefinition = Mock(ServiceDefinition)
+        serviceDefinition.isPresent() >> true
+        serviceDefinition.load() >> new YamlPropertySourceLoader()
+
+        Environment env = new DefaultEnvironment(["test"] as String[]) {
+            @Override
+            protected SoftServiceLoader<PropertySourceLoader> readPropertySourceLoaders() {
+                GroovyClassLoader gcl = new GroovyClassLoader()
+                gcl.addClass(YamlPropertySourceLoader)
+                gcl.addURL(YamlPropertySourceLoader.getResource("/META-INF/services/io.micronaut.context.env.PropertySourceLoader"))
+                return new SoftServiceLoader<PropertySourceLoader>(PropertySourceLoader, gcl)
+            }
+
+            @Override
+            Optional<InputStream> getResourceAsStream(String path) {
+                if(path.endsWith('-test.yml')) {
+                    return Optional.of(new ByteArrayInputStream('''\
+datasources.default: {}
+'''.bytes))
+                }
+                else if(path.endsWith("application.yml")) {
+                    return Optional.of(new ByteArrayInputStream('''\
+datasources.default: {}    
+'''.bytes))
+                }
+
+                return Optional.empty()
+            }
+
+        }
+
+
+        when:
+        env.start()
+
+        then:
+        env.get("datasources.default", String).get() == "{}"
+        env.get("datasources.default", Map).get() == [:]
+
+    }
 }

